File: services/ai/gateway/internal/queue/manager.go
package queue

import (
	"container/heap"
	"context"
	"fmt"
	"sync"
	"time"

	"github.com/rs/zerolog"

	"github.com/stocktracker/gateway/internal/config"
)

// Manager manages the priority queue with CLI concurrency control
type Manager struct {
	config        *config.Config
	logger        zerolog.Logger
	mu            sync.Mutex
	pq            priorityQueue
	running       int // currently executing CLI processes
	maxConcurrent int
	cond          *sync.Cond
	stopped       bool
}

// NewManager creates a new queue manager
func NewManager(cfg *config.Config, logger zerolog.Logger) *Manager {
	m := &Manager{
		config:        cfg,
		logger:        logger,
		pq:            make(priorityQueue, 0),
		maxConcurrent: cfg.MaxConcurrent,
	}
	m.cond = sync.NewCond(&m.mu)
	heap.Init(&m.pq)
	return m
}

// Start begins the queue processor
func (m *Manager) Start(ctx context.Context) {
	go func() {
		<-ctx.Done()
		m.Stop()
	}()
	m.logger.Info().Int("max_concurrent", m.maxConcurrent).Msg("Queue manager started")
}

// Stop signals the queue to stop
func (m *Manager) Stop() {
	m.mu.Lock()
	m.stopped = true
	m.cond.Broadcast()
	m.mu.Unlock()
	m.logger.Info().Msg("Queue manager stopped")
}

// Enqueue adds a request to the priority queue and waits for a CLI slot
// Returns a release function to call when done
func (m *Manager) Enqueue(ctx context.Context, tier config.Tier) (func(), error) {
	tierCfg := m.config.GetTierConfig(tier)

	m.mu.Lock()

	// Check queue depth for this tier
	tierCount := 0
	for _, item := range m.pq {
		if item.Tier == tier {
			tierCount++
		}
	}
	if tierCount >= tierCfg.QueueDepth {
		m.mu.Unlock()
		return nil, fmt.Errorf("queue full for tier %s", tier)
	}

	// Add to priority queue
	item := &Item{
		Tier:      tier,
		Priority:  tierCfg.QueuePriority,
		Timestamp: time.Now(),
		Ready:     make(chan struct{}),
	}
	heap.Push(&m.pq, item)
	m.mu.Unlock()

	// Try to dispatch immediately
	m.tryDispatch()

	// Wait for our turn or timeout
	select {
	case <-item.Ready:
		// We got a slot!
		release := func() {
			m.mu.Lock()
			m.running--
			m.mu.Unlock()
			m.tryDispatch()
		}
		return release, nil
	case <-ctx.Done():
		// Context cancelled/timeout
		m.removeItem(item)
		return nil, ctx.Err()
	case <-time.After(60 * time.Second):
		// Queue wait timeout
		m.removeItem(item)
		return nil, fmt.Errorf("queue wait timeout")
	}
}

// tryDispatch checks if we can dispatch the next item
func (m *Manager) tryDispatch() {
	m.mu.Lock()
	defer m.mu.Unlock()

	for m.running < m.maxConcurrent && m.pq.Len() > 0 {
		item := heap.Pop(&m.pq).(*Item)
		m.running++
		close(item.Ready) // Signal that this item can proceed
	}
}

// removeItem removes an item from the queue (for cancellation)
func (m *Manager) removeItem(item *Item) {
	m.mu.Lock()
	defer m.mu.Unlock()

	for i, qi := range m.pq {
		if qi == item {
			heap.Remove(&m.pq, i)
			break
		}
	}
}

// Stats returns current queue statistics
type Stats struct {
	QueueDepth    int `json:"queue_depth"`
	Running       int `json:"running"`
	MaxConcurrent int `json:"max_concurrent"`
}

func (m *Manager) Stats() Stats {
	m.mu.Lock()
	defer m.mu.Unlock()
	return Stats{
		QueueDepth:    m.pq.Len(),
		Running:       m.running,
		MaxConcurrent: m.maxConcurrent,
	}
}

// priorityQueue implements heap.Interface
type priorityQueue []*Item

func (pq priorityQueue) Len() int { return len(pq) }

func (pq priorityQueue) Less(i, j int) bool {
	// Higher priority first
	if pq[i].Priority != pq[j].Priority {
		return pq[i].Priority > pq[j].Priority
	}
	// Same priority: FIFO (earlier timestamp first)
	return pq[i].Timestamp.Before(pq[j].Timestamp)
}

func (pq priorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}

func (pq *priorityQueue) Push(x interface{}) {
	item := x.(*Item)
	item.index = len(*pq)
	*pq = append(*pq, item)
}

func (pq *priorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	old[n-1] = nil
	item.index = -1
	*pq = old[:n-1]
	return item
}

File: services/ai/gateway-2.0/src/core/queue/manager.ts
/**
 * Priority queue manager for gateway request scheduling.
 *
 * Ported from the Go implementation that uses container/heap + sync.Mutex.
 * In Node.js (single-threaded), we replace mutexes with synchronous code
 * and Go channels with Promises.
 */

import { Heap } from "heap-js";
import type { FastifyBaseLogger } from "fastify";
import type { GatewayConfig } from "../../config.js";
import { type Tier, getTierConfig } from "../../config.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

interface QueueItem {
  tier: Tier;
  priority: number;
  timestamp: number; // Date.now()
  resolve: () => void;
  reject: (reason: Error) => void;
  settled: boolean;
}

export interface QueueStats {
  queueDepth: number;
  running: number;
  maxConcurrent: number;
}

// ---------------------------------------------------------------------------
// QueueManager
// ---------------------------------------------------------------------------

export class QueueManager {
  private readonly heap: Heap<QueueItem>;
  private readonly maxConcurrent: number;
  private readonly logger: FastifyBaseLogger;
  private running = 0;
  private stopped = false;

  constructor(config: GatewayConfig, logger: FastifyBaseLogger) {
    this.maxConcurrent = config.maxConcurrent;
    this.logger = logger;

    // Higher priority first (descending); ties broken by earliest timestamp (FIFO).
    this.heap = new Heap<QueueItem>((a, b) => {
      if (a.priority !== b.priority) return b.priority - a.priority;
      return a.timestamp - b.timestamp;
    });
  }

  // -------------------------------------------------------------------------
  // enqueue
  // -------------------------------------------------------------------------

  /**
   * Enqueue a request for a given tier. Resolves once a concurrency slot is
   * available. The returned function **must** be called to release the slot.
   *
   * @param tier    - The subscription tier of the caller.
   * @param signal  - Optional AbortSignal for cancellation.
   * @returns A release function that frees the concurrency slot.
   */
  async enqueue(tier: Tier, signal?: AbortSignal): Promise<() => void> {
    if (this.stopped) {
      throw new Error("queue is stopped");
    }

    const tierConfig = getTierConfig(tier);

    // Check per-tier queue depth.
    const tierDepth = this.heap
      .toArray()
      .filter((item) => item.tier === tier).length;

    if (tierDepth >= tierConfig.maxQueueDepth) {
      throw new Error(`queue full for tier ${tier}`);
    }

    // Build the queue item with a Promise that resolves when dispatched.
    let itemResolve!: () => void;
    let itemReject!: (reason: Error) => void;

    const gate = new Promise<void>((resolve, reject) => {
      itemResolve = resolve;
      itemReject = reject;
    });

    const item: QueueItem = {
      tier,
      priority: tierConfig.priority,
      timestamp: Date.now(),
      resolve: itemResolve,
      reject: itemReject,
      settled: false,
    };

    this.heap.push(item);
    this.logger.debug(
      { tier, priority: item.priority, queueDepth: this.heap.size() },
      "enqueued request",
    );

    // Attempt to dispatch immediately (synchronous).
    this.tryDispatch();

    // 60-second timeout guard.
    const timeout = setTimeout(() => {
      if (!item.settled) {
        item.settled = true;
        this.heap.remove(item);
        item.reject(new Error("queue timeout"));
      }
    }, 60_000);

    // Optional abort signal handling.
    const onAbort = (): void => {
      if (!item.settled) {
        item.settled = true;
        this.heap.remove(item);
        item.reject(new Error("queue aborted"));
      }
    };

    if (signal) {
      if (signal.aborted) {
        // Already aborted before we even enqueued – clean up immediately.
        if (!item.settled) {
          item.settled = true;
          this.heap.remove(item);
          clearTimeout(timeout);
          throw new Error("queue aborted");
        }
      }
      signal.addEventListener("abort", onAbort, { once: true });
    }

    try {
      await gate;
    } finally {
      clearTimeout(timeout);
      if (signal) {
        signal.removeEventListener("abort", onAbort);
      }
    }

    // Return an idempotent release function.
    let released = false;
    return () => {
      if (released) return;
      released = true;
      this.running--;
      this.logger.debug(
        { tier, running: this.running },
        "released concurrency slot",
      );
      this.tryDispatch();
    };
  }

  // -------------------------------------------------------------------------
  // tryDispatch
  // -------------------------------------------------------------------------

  /**
   * Dispatch as many queued items as available concurrency slots allow.
   * Because Node.js is single-threaded this is purely synchronous.
   */
  private tryDispatch(): void {
    while (this.running < this.maxConcurrent && this.heap.size() > 0) {
      const item = this.heap.pop();
      if (!item) break;

      // Skip already-settled items (timed-out / aborted).
      if (item.settled) continue;

      item.settled = true;
      this.running++;
      this.logger.debug(
        { tier: item.tier, running: this.running },
        "dispatched request",
      );
      item.resolve();
    }
  }

  // -------------------------------------------------------------------------
  // stats
  // -------------------------------------------------------------------------

  stats(): QueueStats {
    return {
      queueDepth: this.heap.size(),
      running: this.running,
      maxConcurrent: this.maxConcurrent,
    };
  }

  // -------------------------------------------------------------------------
  // stop
  // -------------------------------------------------------------------------

  /**
   * Gracefully stop the queue, rejecting all pending items.
   */
  stop(): void {
    this.stopped = true;
    const pending = this.heap.toArray();
    for (const item of pending) {
      if (!item.settled) {
        item.settled = true;
        item.reject(new Error("queue stopped"));
      }
    }
    this.heap.clear();
    this.logger.info("queue manager stopped");
  }
}
